lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+  isBrandCheck('ReadableStream');
+const isWritableStream =
+  isBrandCheck('WritableStream');
+const isTransformStream =
+  isBrandCheck('TransformStream');
+const isReadableByteStreamController =
+  isBrandCheck('ReadableByteStreamController');
+const isReadableStreamBYOBRequest =
+  isBrandCheck('ReadableStreamBYOBRequest');
+const isReadableStreamDefaultReader =
+  isBrandCheck('ReadableStreamDefaultReader');
+const isReadableStreamBYOBReader =
+  isBrandCheck('ReadableStreamBYOBReader');
+const isWritableStreamDefaultWriter =
+  isBrandCheck('WritableStreamDefaultWriter');
+const isWritableStreamDefaultController =
+  isBrandCheck('WritableStreamDefaultController');
+const isTransformStreamDefaultController =
+  isBrandCheck('TransformStreamDefaultController');
+
+function internalExtend(ctor, actual) {
+  ctor.prototype.constructor = actual;
+  ObjectSetPrototypeOf(ctor.prototype, actual.prototype);
+  ObjectDefineProperties(
+    ctor.prototype,
+    ObjectGetOwnPropertyDescriptors(actual.prototype));
+}
+
+function transferArrayBuffer(buffer) {
+  const res = detachArrayBuffer(buffer);
+  if (res === undefined) {
+    throw new ERR_INVALID_STATE.TypeError(
+      'Given ArrayBuffer is not detachable');
+  }
+  return res;
+}
+
+function setPromiseHandled(promise) {
+  // Alternatively, we could use the native API
+  // MarkAsHandled, but this avoids the extra boundary cross
+  // and is hopefully faster at the cost of an extra Promise
+  // allocation.
+  PromisePrototypeThen(promise, () => {}, () => {});
+}
+
+function dequeueValue(controller) {
+  assert(controller[kState].queue !== undefined);
+  assert(controller[kState].queueTotalSize !== undefined);
+  assert(controller[kState].queue.length);
+  const {
+    value,
+    size,
+  } = ArrayPrototypeShift(controller[kState].queue);
+  controller[kState].queueTotalSize =
+    MathMax(0, controller[kState].queueTotalSize - size);
+  return value;
+}
+
+function resetQueue(controller) {
+  assert(controller[kState].queue !== undefined);
+  assert(controller[kState].queueTotalSize !== undefined);
+  controller[kState].queue = [];
+  controller[kState].queueTotalSize = 0;
+}
+
+function peekQueueValue(controller) {
+  assert(controller[kState].queue !== undefined);
+  assert(controller[kState].queueTotalSize !== undefined);
+  assert(controller[kState].queue.length);
+  return controller[kState].queue[0].value;
+}
+
+function enqueueValueWithSize(controller, value, size) {
+  assert(controller[kState].queue !== undefined);
+  assert(controller[kState].queueTotalSize !== undefined);
+  size = Number(size);
+  if (typeof size !== 'number' ||
+      size < 0 ||
+      NumberIsNaN(size) ||
+      size === Infinity) {
+    throw new ERR_INVALID_ARG_VALUE.RangeError('size', size);
+  }
+  ArrayPrototypePush(controller[kState].queue, { value, size });
+  controller[kState].queueTotalSize += size;
+}
+
+// ---- ReadableStream Implementation
+
+function readableStreamPipeTo(
```

Comment:
should this also be an async function? it sometimes returns a rejected promise, and sometimes undefined.

Reply:
That would seem safer, yes.

We can do the same with `ReadableStream.pipeTo()`, and remove the manual `try { } catch (error) { return PromiseReject(error) }` there.

Reply:
So again, yes, but doing so breaks the web platform test that checks a patched Promise.prototype.then ...